### PR TITLE
enh(multi-actions): get generation from `multi-actions-agent`

### LINKED ISSUE #5006: [Multi-Actions] Generation should be done via `use-tools` directly
We have tried to add a `reply_to_user` tool, hoping that the model would use it. We currently call the `generator` action after the actions loop.
Instead, we want to remove instructions that nudges the model away from replying directly, remove the `reply_to_user` tool we are adding, and let the model generate directly from the `use-tools` model call. This needs to be handled properly by the mult-actions loop and treated as the generation (should stop the loop)

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

fixes https://github.com/dust-tt/dust/issues/5006

- rename `use-tools` into `multi-actions-agent`
- fixes a few multi-actions bugs- uses `use`
- use the generation from `mulit-actions-agent` 
- implement streaming the generation from the `multi-actions-agent` action

## Risk

Shouldn't affect single-action agents

### CODE REVIEW

**Comment by spolu** [comment]
We can likely rename the app by now :)

**Comment by fontanierh** [comment]
How would you name it @spolu  ?

**Comment by spolu** [comment]
"agent" ?

**Comment by fontanierh** [comment]
just "agent" ?

**Comment by spolu** [comment]
Well as per our nomenclature that would be assitant-v2-agent?

**Comment by spolu** [comment]
or maybe asssitant-v2-mutli-actions ? or assistant-v2-mutli-actions-agent

File: types/src/front/lib/api/assistant/agent.ts [hunk 1]
```diff
@@ -83,3 +87,11 @@ export type AgentMessageSuccessEvent = {
   messageId: string;
   message: AgentMessageType;
 };
+
+export type AgentUseToolEvent = {
```
[the remark this answers is not in the document]
Comment (spolu) [reply types/src/front/lib/api/assistant/agent.ts:91]: If we don't need to introduce a new name we definitely shouldn't and it looks like we don't really need it ?

File: front/pages/api/w/[wId]/assistant/agent_configurations/index.ts [hunk 2]
```diff
@@ -222,19 +222,6 @@ async function handler(
         });
       }
 
-      if (
```
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/agent_configurations/index.ts:225]: Why remove this?

File: front/lib/api/assistant/agent.ts [hunk 3]
```diff
@@ -398,66 +371,204 @@ export async function getNextAction(
...
+    generation: null,
+  };
+
+  let shouldYieldCancel = false;
+  let lastCheckCancellation = Date.now();
+  const redis = await redisClient();
+  let isGeneration = true;
+
+  try {
+    const _checkCancellation = async () => {
```
Comment (spolu) [inline front/lib/api/assistant/agent.ts:405]: What's this?

File: front/lib/api/assistant/agent.ts [hunk 4]
```diff
@@ -398,66 +371,204 @@ export async function getNextAction(
...
+      if (event.type === "block_execution") {
+        const e = event.content.execution[0][0];
+        if (e.error) {
+          yield {
+            type: "agent_error",
+            created: Date.now(),
+            configurationId: agentConfiguration.sId,
+            messageId: agentMessage.sId,
+            error: {
+              code: "use_tools_error",
```
Comment (spolu) [inline front/lib/api/assistant/agent.ts:487]: Let's not use `tools` here either

Comment (spolu) [reply front/lib/api/assistant/agent.ts:487]: Maybe `mutli_actions_error`

**Review by spolu — COMMENTED** [review]
Left suggestions and comments :pray:

File: types/src/front/lib/api/assistant/agent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fontanierh) [reply types/src/front/lib/api/assistant/agent.ts:91]: Yep totally agree

File: front/pages/api/w/[wId]/assistant/agent_configurations/index.ts (hunk 2 above)
Comment (fontanierh) [reply front/pages/api/w/[wId]/assistant/agent_configurations/index.ts:225]: Screenshot 2024-05-13 at 20 33 39


🤦

File: front/lib/api/assistant/agent.ts (hunk 3 above)
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:405]: same as in runGeneration, allows to react to cancellation from UI (it's cargo-culted from there)

File: types/src/front/lib/api/assistant/agent.ts [hunk 5]
```diff
@@ -41,7 +45,7 @@ export type AgentErrorEvent = {
 };
 
 // Event sent during the execution of an action. These are action specific.
-export type AgentActionEvent =
+export type AgentActionParamsOrOutputEvent =
```
Comment (spolu) [inline types/src/front/lib/api/assistant/agent.ts:48]: Maybe AgentActionSpecificEvent?

**Review by spolu — APPROVED** [review]

File: types/src/front/lib/api/assistant/agent.ts (hunk 5 above)
Comment (fontanierh) [reply types/src/front/lib/api/assistant/agent.ts:48]: I'm not sure I fully understand the suggestion, but I am fine with it 👍

**Pull request merged**